vignettes/design_principle.Rmd

```diff
@@ -0,0 +1,90 @@
+
+To address this challenge, PIs of the [Epiverse-TRACE](https://data.org/initiatives/epiverse/) came with a list of potential data sources for which we aim at building a tool to make requests and fetch the data of interest in the same way. The data sources include: distributed health information systems, and public databases as shown in the figure below.
+
+![readepi roadmap](../man/figures/roadmap-readepi.drawio.png)
+
+## Scope
+
+The {readepi} package intends to import data from two common sources of institutional health-related data: health information systems (HIS) wrapped with specific program interfaces (APIs) and relational database management systems (RDBMS) that run on specific servers. Each source has its own import function, `read_from_api()` for HIS and `read_from_server()` for RDBMS. There is also an auxiliary function, `get_metadata()`, that retrieves metadata from a specified API and offers interoperability with the [{cleanepi} R package](https://epiverse-trace.github.io/cleanepi/).     
+The current version of {readepi} supports importing data from HIS APIs such as REDCap (Research Electronic Data Capture), DHIS2 (District Health Information System 2), Fingertips, GoData, and Globaldothealth, as well as RDBMS such as MS SQL, MS ACCESS, SQlite, MYSQL, and PostgreSQL. 
+
+![readepi design](../man/figures/readepi_design.svg)
+
+
+## Output
+
+The two import functions return a `list` object containing one or more `data frames`. Each `data frame` corresponds to the data from a specified table or file. The `get_metadata()` function returns a data dictionary containing information about the data structure. 
+
+## Design decisions
+
+The aim of {readepi} is to simplify and standardize the process of fetching health data from APIs and servers. We ambition to make it requires minimal information to access and pull the data of interest from the target source.  
+
+1. The `read_from_api()` function takes only two arguments to import data from an API: an object of class **credential** and an object of class **api**.
+
+The **credential** object is a list-like object that stores the information required to access the API. Its main attributes are:
+    * `base_url`: the link to the API,
+    * `user_name`: the user name, and
+    * `password`: the user's password or API token.
+
+The **api** object is also a list-like object that provides information about the API. It has 2 main attributes:
+    * `name`: the API name
+    * `attributes`: the attributes needed to access the data of interest
+Note that the API's attributes can vary depending on the API, for example, it can be dataElements or dataSets or a combination of both for DHIS2. 
+Both the API name and attributes are needed to build the API endpoint.
```

Comment:
Is there any further opportunity for harmonization here or is it unavoidable that `attributes` will be very different from one case to the other?

If it can take very different values, how to document it clearly?

Reply:
It is challenging to harmonize the attributes because of the wide variations in the internal structures of the APIs. In separate paragraphs, we will offer a thorough documentation of the properties for each API that is being considered.